src/refract/JSONSchemaVisitor.cc

```diff
@@ -7,646 +7,664 @@
-            str = TypeQueryVisitor::as<StringElement>(merged);
+    auto keyStr = TypeQueryVisitor::as<const StringElement>(key);
+    auto keyExt = TypeQueryVisitor::as<const ExtendElement>(key);
 
-            if (!str) {
-                delete merged;
-            }
+    auto keyMerged = [](const ExtendElement* ext) -> std::unique_ptr<StringElement> {
+        if (ext) {
+            auto merged = ext->get().merge();
+            if (StringElement* ownPtr = TypeQueryVisitor::as<StringElement>(merged.release()))
+                return std::unique_ptr<StringElement>(ownPtr);
         }
+        return nullptr;
+    }(keyExt);
 
-        if (str) {
-            IElement* desc = GetDescription(e);
+    if (keyMerged)
+        keyStr = keyMerged.get();
 
-            if (desc) {
-                IElement* d = desc->clone();
-                renderer.addMember("description", d);
-            }
+    if (keyStr) {
+        const auto& desc = GetDescription(e);
 
-            if (IsTypeAttribute(e, "nullable")) {
-                renderer.addSchemaType("null");
-                renderer.addNullToEnum();
-            }
+        if (desc) {
+            renderer.addMember("description", clone(*desc));
+        }
+
+        if (IsTypeAttribute(e, "nullable")) {
+            renderer.addSchemaType("null");
+            renderer.addNullToEnum();
+        }
 
-            // Check for primitive types
-            StringElement* strSecond = TypeQueryVisitor::as<StringElement>(e.value.second);
-            NumberElement* numSecond = TypeQueryVisitor::as<NumberElement>(e.value.second);
-            BooleanElement* boolSecond = TypeQueryVisitor::as<BooleanElement>(e.value.second);
+        // Check for primitive types
+        const auto& valueStr = TypeQueryVisitor::as<const StringElement>(value);
+        const auto& valueNum = TypeQueryVisitor::as<const NumberElement>(value);
+        const auto& valueBool = TypeQueryVisitor::as<const BooleanElement>(value);
 
-            if (e.value.second && (strSecond || numSecond || boolSecond)) {
-                auto defaultIt = e.value.second->attributes.find("default");
+        if (value && (valueStr || valueNum || valueBool)) {
+            auto defaultIt = value->attributes().find("default");
 
-                if (defaultIt != e.value.second->attributes.end()) {
-                    renderer.addMember("default", (*defaultIt)->clone());
-                }
+            if (defaultIt != value->attributes().end()) {
+                renderer.addMember("default", defaultIt->second->clone());
             }
-
-            addMember(str->value, renderer.getOwnership());
-        } else {
-            throw LogicError("A property's key in the object is not of type string");
         }
 
-        if (ext && str) {
-            delete str; // if Extend -> remove resolved (merged) element
-        }
+        addMember(keyStr->empty() ? std::string{} : std::string{ keyStr->get() }, renderer.getOwnership());
+    } else {
```

Comment:
What about remove this condition and replace it by `assert(keyStr)` in precondition